### PR TITLE
Adds merge command

### PR DESCRIPTION
**Pull request by henriquetruta**
Adds command that, given source and target kubeconfig
files, merge the contents of a source into the target.

Fixes https://github.com/kubernetes/kubernetes/issues/46381 

**Release note**:

```release-note
kubectl merge command - merge the contents of a given source kubeconfig file into a given target file.
```

### CODE REVIEW

**Comment by kargakis** [comment]
cc: @kubernetes/sig-api-machinery-pr-reviews

File: pkg/kubectl/cmd/config/merge_configs.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+var (
+	mergeConfigsLong = templates.LongDesc(`
+		Merges kubecontext files
+
+		Adds all cluster, user and context entries of a source kubeconfig
+		into another kubeconfig target file. If target is not provided,
+		the kubeconfig used in path is considered. In case there is a
+		name conflict, i.e. an entry with a given name exists in both
+		files, this entry is ignored.`)
```
Comment (sttts) [inline pkg/kubectl/cmd/config/merge_configs.go]: s/is ignored/of the target file is preserved/

File: pkg/kubectl/cmd/config/merge_configs.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+// MergeConfigsOptions contains the assignable options from the args.
+type MergeConfigsOptions struct {
+	configAccess clientcmd.ConfigAccess
+	source       string
+	target       string
+}
+
+var (
+	mergeConfigsLong = templates.LongDesc(`
+		Merges kubecontext files
```
Comment (sttts) [inline pkg/kubectl/cmd/config/merge_configs.go]: <bikeshed>`merge` sounds symmetric. `import` would probably be more appropriate</bikeshed>.

File: pkg/kubectl/cmd/config/merge_configs.go (hunk 1 above)
Comment (henriquetruta) [inline pkg/kubectl/cmd/config/merge_configs.go]: done

File: pkg/kubectl/cmd/config/merge_configs.go (hunk 2 above)
Comment (henriquetruta) [inline pkg/kubectl/cmd/config/merge_configs.go]: liked the suggestion. Will apply. Thanks

**Comment by mengqiy** [comment]
/assign @shiywang @alexandercampbell 
Do you have bandwidth to review this?

**Comment by alexandercampbell** [comment]
@mengqiy yep, I'm doing CRs this afternoon.

File: pkg/kubectl/cmd/config/import_configs.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	if err := clientcmd.ModifyConfig(o.configAccess, *config, true); err != nil {
+		return err
+	}
+
+	allErrs := []error{}
+	return utilerrors.NewAggregate(allErrs)
+}
+
+func isEmptyConfig(config *clientcmdapi.Config) bool {
+	return (len(config.Clusters) == 0) && len(config.Contexts) == 0 && len(config.AuthInfos) == 0
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs.go]: You've got some redundant parens here on the first len check.

File: pkg/kubectl/cmd/config/import_configs_test.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+		startingConfig:   tconf,
+		sourceConfig:     source,
+		createSourceFile: true,
+		target:           "",
+		expectedOut: []string{
+			"There is no context entry in the source file.\n",
+			"There is no user entry in the source file.\n",
+			fmt.Sprintf("Copied cluster entry %q to target file.\n", "minikube"),
+			fmt.Sprintf("Copied cluster entry %q to target file.\n", "bigkube"),
+			fmt.Sprintf("Copied cluster entry %q to target file.", "federatedkube")},
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs_test.go]: Why not just put the string literal in here?

File: pkg/kubectl/cmd/config/import_configs_test.go [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+		for _, o := range test.expectedOut {
+			if !strings.Contains(buf.String(), o) {
+				t.Errorf("could not find %q phrase in %q", o, buf.String())
+			}
+		}
+		return
+	}
+	if len(test.expectedErr) != 0 {
+		if err == nil {
+			t.Errorf("Did not get %v", test.expectedErr)
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs_test.go]: I might clarify this string: `"Expected error %v but no error was returned."`

File: pkg/kubectl/cmd/config/import_configs_test.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+	options.Complete(cmd, args, buf)
+	err = options.RunImportConfigs(buf)
+
+	if len(test.expectedOut) != 0 {
+		for _, o := range test.expectedOut {
+			if !strings.Contains(buf.String(), o) {
+				t.Errorf("could not find %q phrase in %q", o, buf.String())
+			}
+		}
+		return
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs_test.go]: Why the return?

File: pkg/kubectl/cmd/config/import_configs.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+		Use:     "import source [(-t|--target)=~/.kube/config]",
+		Short:   i18n.T("Imports kubeconfig file content"),
+		Long:    importConfigsLong,
+		Example: importConfigsExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.Complete(cmd, args, out))
+			cmdutil.CheckErr(options.RunImportConfigs(out))
+		},
+	}
+	cmd.Flags().String("target", "", "the file that will receive the entries. Defaults to the regular kubeconfig file")
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs.go]: If this defaults to the "regular kubeconfig" file, then why is the default a blank string?

File: pkg/kubectl/cmd/config/import_configs.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+		cmd.Help()
+		return fmt.Errorf("Unexpected args: %v", args)
+	}
+	o.source = args[0]
+
+	return nil
+}
+
+// RunImportConfigs implements all the necessary functionality for merging two kubeconfig files.
+func (o ImportConfigsOptions) RunImportConfigs(out io.Writer) error {
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs.go]: I think this method is just supposed to be called "Run"

File: pkg/kubectl/cmd/config/import_configs.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+		},
+	}
+	cmd.Flags().String("target", "", "the file that will receive the entries. Defaults to the regular kubeconfig file")
+	return cmd
+}
+
+// Complete assigns ImportConfigsOptions from the args.
+func (o *ImportConfigsOptions) Complete(cmd *cobra.Command, args []string, out io.Writer) error {
+
+	o.target = ""
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs.go]: You don't need this line.

File: pkg/kubectl/cmd/config/import_configs.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+	if cmdutil.GetFlagString(cmd, "target") != "" {
+		o.target = cmdutil.GetFlagString(cmd, "target")
+	}
+
+	if len(args) != 1 {
+		cmd.Help()
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs.go]: I don't think help should be printed here. It's inconsistent with the other commands.

File: pkg/kubectl/cmd/config/import_configs.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+		return fmt.Errorf("Unexpected args: %v", args)
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs.go]: I think this error message might be clearer if it read: `"import source expects only one argument"`

File: pkg/kubectl/cmd/config/import_configs_test.go [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+	"strings"
+	"testing"
+
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+const (
+	notExistingFile   = "notexisting"
+	targetContextName = "target-context"
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs_test.go]: Might be worth throwing quick comments on these constants.

File: pkg/kubectl/cmd/config/import_configs_test.go [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+			fmt.Sprintf("Cluster called %q exists in both files, so it will not be copied.", "minikube"),
+			fmt.Sprintf("Copied context entry %q to target file.\n", "shaker-context"),
+			fmt.Sprintf("Context called %q exists in both files, so it will not be copied.", "other-context"),
+			fmt.Sprintf("Copied User entry %q to target file.\n", "myuser"),
+			fmt.Sprintf("User called %q exists in both files, so it will not be copied.", "admin")},
+		expectedErr: ""}
+	test.run(t)
+}
+
+func (test importConfigsTest) run(t *testing.T) {
```
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs_test.go]: I would move this method definition next to the struct definition.

**Review by alexandercampbell — CHANGES_REQUESTED** [review]
The code looks mostly fine. My biggest concern is: is it a good idea to add this feature in this way? I'll defer to @bgrant0607.

**Comment by lavalamp** [comment]
> My biggest concern is: is it a good idea to add this feature in this way? I'll defer to @bgrant0607.

@pwittrock Do you think this is a good idea?

**Comment by fabianofranz** [comment]
I didn't check here how you decided to implement the solution for naming conflicts, but I have the same concerns expressed [here](https://github.com/kubernetes/kubernetes/pull/38861#issuecomment-273865070). 

The outcome of running a command like this on sets of large config files is hard to understand/predict so I suggest to adopt the most conservative approach possible. E.g. in case of any naming conflict or "suspicious" merging (e.g a context with cluster from one file and credentials from another), error out and do nothing.

**Review by henriquetruta — COMMENTED** [review]
@fabianofranz thanks for the comment. I haven't considered the "suspicious" merging. I'll treat it in my next commit. Don't you think it's worth to merge all 'safe' entries? i.e. ignore the ones with any issue and apply the ones that are safe. Not the most conservative approach, but seems enough for me

**Comment by liggitt** [comment]
is this the equivalent of `kubectl config view --raw` with multiple kubeconfig files specified?

`KUBECONFIG=file1:file2 kubectl config view --raw`

**Comment by henriquetruta** [comment]
@liggitt It is similar, except that I don't expect that `view --raw` deals with naming conflicts somehow

**Comment by liggitt** [comment]
dealing with naming conflicts differently than the merging logic that is actually applied when you combine files doesn't seem like a good idea

**Comment by henriquetruta** [comment]
@liggitt just made some tests here. Copied a file into kube1 and kube2 files. Then, changed some entries values on kube2, like cluster IP and user passwords and ran:
```
export KUBECONFIG=kube1:kube2
k config view [--raw]
```

In this case, I see only the values of kube1. A similar behaviour to the one I adopted

**Comment by liggitt** [comment]
> A similar behaviour to the one I adopted

Similar, but implemented separately. We should have a single implementation.

File: pkg/kubectl/cmd/config/import_configs.go (hunk 8 above)
Comment (alexandercampbell) [inline pkg/kubectl/cmd/config/import_configs.go]: Gotcha. Thanks.

**Pull request closed without merging**